oat-sa/tao-core
# feat: remove keys call to avoid locking keyset

https://oat-sa.atlassian.net/browse/ADF-1328
https://oat-sa.atlassian.net/browse/ADF-1329

# Goal

- Improve REDIS performance by removing the `keys` call that locked the keyset and is not recommended to use on production (More details here https://oat-sa.atlassian.net/wiki/spaces/OAT/pages/1972731905/RFE-396+Token+Pool+Creation+Optimization#ISSUE-3---Redis-keys).

## Changes done

- Remove the `keys` usage.
- Use hash map to save tokens, saving storage and improving data retrieval from Redis.
- Create a command to delete expired tokens. It needs to be added to a CRON Job.

## How to test

Follow instructions https://oat-sa.atlassian.net/wiki/spaces/OAT/pages/1991311625/How+to+benchmark+TokenService+with+Redis

Also:

- Check the CSRF token validation will work by submitting items as TestTaker.
- ❗ The CSRF token validation inside the Authoring NEVER worked. It only logs a warning and nothing more.
- Run this command to check that tokens created after 120sec are removed `php index.php 'oat\tao\scripts\tools\Security\ClearCsrfTokenTool' -t 120`

## Benchmark results 

The benchmark shows that we have more Network, Ops/sec because of the new code comparing to the previous, but we are not blocking Redis operations for anything else since we removed the usage of `keys`:

Old implementation Vs New

![Screenshot 2023-03-13 at 14 37 19](https://user-images.githubusercontent.com/1467589/224718183-86b3aa43-5a47-499a-a27e-eb47c8b9b942.png)

- Reduced Ops/sec from **2k/s** to **600/s**.
- Reduce input from 150kb to 50kb (3x less).

## Related PRs

- https://github.com/oat-sa/generis/pull/1030 (needed to check)
- https://github.com/oat-sa/tao-core/pull/3442/files (this PR caused the issue with `keys`)

## TODO 

- [x] Fix code styles
